### DISCUSSION TITLE
PeriodicCycleMiner argument passing and covering of new data

### DISCUSSION THREAD

**Discussion by remiadon** [scikit-mine/scikit-mine#123] [Ideas]
@nurblageij for now PeriodicCycleMiner takes a `pd.Series` as argument, with an index containing occurrences, of type `pd.DatetimeIndex | pd.RangeIndex | pd.Int64Index`, and values containing events, of type `object|string`

This translates into the following syntax
```python
>> from skmine.periodic import PeriodicCycleMiner
>> S = pd.Series("ring_a_bell", [10, 20, 32, 40, 60, 79, 100, 240])
>> pcm = PeriodicCycleMiner().fit(S)
>> pcm.discover()
```

## Observations about the logs we pass in
- there can be multiple occurrences, and multiple events
- there is a one-to-one mapping between occurrences and events
- this `pd.Series` "log input" is not handle by any sklearn estimator. In other words, this does not really follows the sklearn API

## Proposition
My proposition is to pass events and occurrences separately, by considering events as "labels"

This would result in the following syntax
```python
>> from skmine.periodic import PeriodicCycleMiner
>> X = [10, 20, 32, 40, 60, 79, 100, 240]
>> y = ["ring_a_bell"] * len(X)
>> pcm = PeriodicCycleMiner().fit(X, y)
>> pcm.discover()
```
### Pros
- This would get us a little bit closer to the sklearn API
- PeriodicCycleMiner would become a "supervised" learning algorithm
- To this regard, we could have a discussion on how to "predict" events given a new set of occurrences, or a least cover them
- One could "evaluate" a PeriodicCycleMiner, on different train/test sets with a given metric, like it is done for ML models

Covering new data is still a process to be defined, but if implemented, one could write something like the following
```python
>> X = [18, 32, 40, 40, 49, 60, 79, 100, 240]
>> y = list("ABABBAAA")
>> pcm = PeriodicCycleMiner().fit(X, y)
>> pcm.discover()
                   start  length  period
              A 0     18       5      20
              B 0     32       3      8

>> pcm.predict([33, 41, 52])  # same length and period close to [32, 40, 49]
["B", "B", "B"]  # all occurrences covered by the same cycle (the second one)
```

### Cons
- The proposed syntax seems a bit less intuitive for users considering the input data as "logs".
- Does this event make sense to cover new data ? What are the practical applications ?

---------------

What do you think ?
